invenio_records_restapi/restful.py

```diff
@@ -0,0 +1,144 @@
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""REST API resources."""
+
+from flask import jsonify, make_response, request, url_for
+
+from flask.ext.restful import Resource, abort
+
+from invenio_db import db
+
+from invenio_records.api import Record
+
+from jsonpatch import JsonPatchException, JsonPointerException
+
+from sqlalchemy.orm.exc import NoResultFound
+
+
+def record_self_link(record, **kwargs):
+    """Creates self link to a given record.
+
+    :Parameters:
+        - `record` (Record): record to which the generated link will point.
+        - `**kwargs`: additional parameters given to flask.url_for.
+    :Returns: link pointing to the given record.
+    :Returns Type: str
+    """
+    # '.' is needed to reference the current blueprint context
+    return url_for('.' + RecordAPI.api_endpoint,
+                   record_id=record.model.id, **kwargs)
+
+
+def request_wants_json():
+    """Test if the request accept type accepts 'application/json'.
+    :Returns: True if the request's accept type contains 'application/json'
+    else False.
+    :Returns Type: bool
+    """
+    best = request.accept_mimetypes \
+        .best_match(['application/json'])
+    return best == 'application/json'
+
+
+def record_to_response(record):
+    """Build a flask response from the given record, formatting it in an
+    accepted format.
+    :Returns: A flask response.
+    :Returns Type: Response
+    """
+    if not request_wants_json():
+        # requested mime format not supported
+        abort(406)
+
+    # FIXME: use a formatter instead once it is implemented
+    self = record_self_link(record, _external=True)
+    response = jsonify({
+        'id': record.model.id,
+        'data': record,
+        'links': {
+            'self': self
+        }
+    })
+    response.headers['location'] = self
+    return response
+
+
+class RecordsRootAPI(Resource):
+    """Resource for '/records'"""
+
+    api_endpoint = 'invenio_records_restapi__records'
+
+    def post(self, **kwargs):
+        """Create a Record.
+        :Returns: The created record.
+        """
+        data = request.get_json()
+        if data is None:
+            return abort(400, status=400, message="Invalid data")
+        record = Record.create(data)
+        db.session.commit()
+        response = record_to_response(record)
+        response.status_code = 201  # created
+        return response
+
+
+class RecordAPI(Resource):
+    """Resource for '/records/<int:record_id>'"""
+
+    api_endpoint = 'invenio_records_restapi__records_item'
+
+    def get(self, record_id, **kwargs):
+        """Get a Record.
+        :Parameters:
+            - `record_id`: id of the record to retrieve.
+        :Returns: The requested record.
+        """
+        try:
```

Comment:
A big part of this code here, will be very similar to code in invenio-records-ui. We should figure out a way that the common parts can be shared.